### ISSUE TITLE
v1.2.0-37a3aa0 (Dec 21, 2018) truncated "tags.data" and recreates only two bytes long file

### ISSUE THREAD

**Issue by sskras** [GothenburgBitFactory/timewarrior#292]
(Long intro. Possibly dup of #265, though I run it on Msys2)

I have a Bash script called `rlmtu` that reports unique time tracks from the last month:
```
$ rlmtu
Reporting time range: 2020-01-03T10:00:07 - 2020-01-31T23:46:52

--- monthly time declaration mark ---, [pro-time]
[bureau_], attend Linos presentation about ethics of the IT worker/admin
[organze], city travel
[organze], go reg working time
  ...
[pro-wiki], extend ADB documentation
[pro-wiki], extend VMS/O3G documentation
[pro-wiki], read the DCIM integration docs
```

Now if I cut the first tag + pass through only tags without the `[` symbol, I get a nice list of tags (with names coming from the TaskWarrior project).
```
$ rlmtu | awk 'BEGIN {FS=","} {print $1}' | sort | uniq | awk '/\[/'
[bureau_]
[organze]
[pro-asst]
  ...
[pro-secr]
[pro-updt]
[pro-wiki]
```

To calculate the monthly time of project, I use `timew summ` + `tail`:
```
$ timew summ 2020-01-03T10:00:07 - 2020-01-31T23:46:52 [organze] | tail
                  [organze], plan tasks                     12:08:35 12:10:57 0:02:22
                  [organze], plan tasks                     13:08:50 13:10:13 0:01:23
                  [organze], register / adjust task names   13:43:09 13:52:13 0:09:04
                  [organze], register / adjust task names   14:03:03 14:04:24 0:01:21
                  [organze], register / adjust task names   14:30:13 14:39:32 0:09:19
                  [organze], plan tasks                     23:28:45 23:37:50 0:09:05
                  [organze], go reg working time            23:37:54 23:46:49 0:08:55  1:20:01

                                                                                      33:57:32
```

Now I put this list into the for-loop in Bash and used `tail -n 1` to list monthly times per project:
```
$ for PROJ in $(rlmtu | awk 'BEGIN {FS=","} {print $1}' | sort | uniq | awk '/\[/'); do \
>     echo -n Processing $PROJ:; \
>     timew summ 2020-01-03T10:00:07 - 2020-01-31T23:46:52 [pro-bckp] | tail -n 1; \
> done
```

Oops, something got broken, it seems.  Also maybe I pressed `^C` at the time.

### **The possible occurrence of the bug.**

After adding a debugging echo I got already different error coming from inside my `rlmtu` script:
```
$ for PROJ in $(rlmtu | awk 'BEGIN {FS=","} {print $1}' | sort | uniq | awk '/\[/'); do \
>     echo -n Processing $PROJ:; \
>     echo timew summ 2020-01-03T10:00:07 - 2020-01-31T23:46:52 [pro-bckp] | tail -n 1; \
> done
Error: expected '{' or '[' at position 0
Error: expected '{' or '[' at position 0
Error: expected '{' or '[' at position 0
Error: expected '{' or '[' at position 0
```

Which traces to:
```
++ timew summ :id :year '--- monthly time declaration mark ---'
Error: expected '{' or '[' at position 0
```

Now I see `tags.data` got truncated : (
```
$ ll /home/saukrs/.timewarrior/data/
total 6.2M
-rw-r--r-- 1 saukrs Domain Users  34K Aug 30 17:54 2019-01.data
-rw-r--r-- 1 saukrs Domain Users  53K Aug 30 17:54 2019-02.data
-rw-r--r-- 1 saukrs Domain Users  45K Aug 14  2019 2019-03.data
-rw-r--r-- 1 saukrs Domain Users  46K Aug 30 17:54 2019-04.data
-rw-r--r-- 1 saukrs Domain Users  22K Aug 21  2019 2019-05.data
-rw-r--r-- 1 saukrs Domain Users  36K Aug 25  2019 2019-06.data
-rw-r--r-- 1 saukrs Domain Users  45K Aug 30 18:02 2019-07.data
-rw-r--r-- 1 saukrs Domain Users  49K Nov 28 12:56 2019-08.data
-rw-r--r-- 1 saukrs Domain Users  46K Nov  2 15:46 2019-09.data
-rw-r--r-- 1 saukrs Domain Users  40K Dec 31 14:15 2019-10.data
-rw-r--r-- 1 saukrs Domain Users  45K Dec 31 14:39 2019-11.data
-rw-r--r-- 1 saukrs Domain Users  36K Dec 31 22:07 2019-12.data
-rw-r--r-- 1 saukrs Domain Users  33K Feb 28 13:48 2020-01.data
-rw-r--r-- 1 saukrs Domain Users  39K Feb 28 21:26 2020-02.data
-rw-r--r-- 1 saukrs Domain Users    0 Feb 28 21:50 tags.data
-rw-r--r-- 1 saukrs Domain Users 5.6M Feb 28 21:26 undo.data
```

There was some disk space available at the time:
```
$ df -Th .
Filesystem               Type  Size  Used Avail Use% Mounted on
C:/Users/saukrs/_/msys64 ntfs  237G  236G  1.2G 100% /
```
---

1, how can I restore the `tags.data` ?
2, how could that happen if I seemingly didn't do any writes to the `~./timewarrior` ?

**Comment by sruffell** [comment]
> 1, how can I restore the tags.data ?

In 1.2.0, you can delete `tags.data` and timewarrior will regenerate it from the intervals in the database.  In the current `dev` branch, if timewarrior fails to parse the tags database, it will recreate it.

> 2, how could that happen if I seemingly didn't do any writes to the ~./timewarrior ?

In 1.2.0, the tags.data file is regenerated every time `timew` is run. This was reported in #284, and is fixed on the current dev branch in commit aa7a674ca695fe .


As to how this occurred, I strongly suspect the problem was when you said "ooops I hit ctrl-c", maybe in the middle of writing the tags database, or between when it was truncated and when it was updated. 

The current dev branch should be less susceptible to these sorts of issues, since temporary files are written first, and are then renamed over the "real" files. Therefore, the window of time where early program termination could leave the database in a "bad" state is smaller (however, reading your issue, `timew` could probably benefit from a signal handler to cleanup temporary files on any termination, and mask SIGINTs during the rename operation.)

In summary, can you try with the current dev branch? I use it every day and believe it's safe, but you may want to ensure you have a backup of your database (but you do back that up regularly anyway, right?)

**Comment by sskras** [comment]
@sruffell, thanks a lot. Quoting:

> > 1, how can I restore the tags.data ?
> 
> In 1.2.0, you can delete `tags.data` and timewarrior will regenerate it from the intervals in the database. 

Sadly, it did not:

```
$ rm /home/saukrs/.timewarrior/data/tags.data
$ timew summ

Wk Date       Day Tags                                                                                     Start End     Time    Total
W9 2020-02-29 Sat [PVw-fixs], subscribe for new LUN to use for direct SD...1 data export-import on H...s 0:00:00   - 20:17:35 20:17:35

                                                                                                                              20:17:35
```

As I got summary, I hoped every everything is fine now and started a task:
```
$ task '/new task/' start
Trying to decrement non-existent tag '[PVw-fixs]'
Starting task 456ea389-1421-4612-aecc-2ceac009f2e8 'register new task'.
Started 1 task.
Project '[organze]' is 4% complete (95 of 99 tasks remaining).
```

But when I tried to edit unstopped task of yesterday, TimeW failed:
```
$ timew modi end @2 2020-02-28T21:55:03
Trying to decrement non-existent tag '[organze]'
```

The `tags.data` was regenerated, but almost empty:
```
$ ll /home/saukrs/.timewarrior/data/tags.data
-rw-r--r-- 1 saukrs Domain Users 2 Feb 29 20:19 /home/saukrs/.timewarrior/data/tags.data

$ cat /home/saukrs/.timewarrior/data/tags.data
{}
```

Maybe it's expected with TimeWarrior v1.2.0 ?

**Comment by sruffell** [comment]
What is the output of `timew diagnostics`?

I'm not sure you're running 1.2.0. Below you can see that when I run `timew summ` after deleting the tags database, timewarrior displays `Tag info database does not exist. Recreating from interval data...` which I did not see in your output.

```
$ rm -fr ~/.timewarrior
$ timew start tag1
Create new database in /home/sruffell/.timewarrior? (yes/no) yes
Note: 'tag1' is a new tag.
Tracking tag1
  Started 2020-02-29T12:42:20
  Current                  20
  Total               0:00:00
$ timew start tag2
Recorded tag1
  Started 2020-02-29T12:42:20
  Ended                    22
  Total               0:00:02
Note: 'tag2' is a new tag.
Tracking tag2
  Started 2020-02-29T12:42:22
  Current                  22
  Total               0:00:00
$ timew summ

Wk Date       Day Tags    Start      End    Time   Total
W9 2020-02-29 Sat tag1 12:42:20 12:42:22 0:00:02
                  tag2 12:42:22        - 0:00:05 0:00:07

                                                 0:00:07

$ rm ~/.timewarrior/data/tags.data
$ timew summ
Tag info database does not exist. Recreating from interval data...

Wk Date       Day Tags    Start      End    Time   Total
W9 2020-02-29 Sat tag1 12:42:20 12:42:22 0:00:02
                  tag2 12:42:22        - 0:00:12 0:00:14

                                                 0:00:14

$ timew diagnostics

timew 1.2.0
       Platform: Linux

Compiler:
        Version: 7.4.0
           Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
     Compliance: C++11

Build Features
          Built: Feb 29 2020 12:41:55
         Commit: 1d271508
          CMake: 3.10.2
     Build type: release

Configuration
  TIMEWARRIORDB: -
            Cfg: /home/sruffell/.timewarrior/timewarrior.cfg (-rw- 0 bytes)
       Database: /home/sruffell/.timewarrior (drwx 4096 bytes)
        $EDITOR: vim
    Color theme: Built-in default
                 00 01 02 03 04 05 06 07 08 09 10 11 12

Extensions
       Location: /home/sruffell/.timewarrior/extensions (drwx 4096 bytes)
               (None)

$ cat ~/.timewarrior/data/tags.data
{"tag1":{"count":1},"tag2":{"count":1}}$
```
```

**Comment by sskras** [comment]
Had forgotten that magic command:
```
$ timew diagnostics

timew 1.2.0
       Platform: <unknown>

Compiler:
        Version: 7.3.0
           Caps: +stdc +stdc_hosted +200809 +200809 +LP64 +c8 +i32 +l64 +vp64 +time_t64
     Compliance: C++11

Build Features
          Built: Apr 15 2019 16:17:24
         Commit: 37a3aa0
          CMake: 3.13.2
     Build type: debug

Configuration
  TIMEWARRIORDB: -
            Cfg: /home/saukrs/.timewarrior/timewarrior.cfg (-rw- 64 bytes)
       Database: /home/saukrs/.timewarrior (drwx 0 bytes)
    Color theme: Built-in default
                 00 01 02 03 04 05 06 07 08 09 10 11 12

Extensions
       Location: /home/saukrs/.timewarrior/extensions (drwx 0 bytes)
               (None)
```

So I built it from commit 37a3aa0, circa Dec 21, 2018.

**Comment by sruffell** [comment]
> So I built it from commit 37a3aa0, circa Dec 21, 2018.

Ahh, so, yeah, you do not have everything that is in the tagged version of 1.2.0, since the commit that regenerates the tags went in around May of 2019.

I recommend running from the current tip of the dev branch if you can.

**Comment by sskras** [comment]
Thanks, but I guess we will need to come back here soon. Postponing this to register new bug for the dev branch.

**Comment by sskras** [comment]
So I tested the b15cddd562760c4bee28534a75a0523efdaad03e (May 30 2019), and it's fixed there.
